Operational mission

christophe-david opened the pull request:
This PR proposes an extension of the YAML-defined mission file for computing missions inside the sizing loop (design mission) and outside the sizing loop (operational missions).
It also unifies the performance computation, by allowing to specify the legacy Breguet algorithm through a YAML mission file.

It also fixes #223.

A documentation for the mission module has also been written:
https://fast-oad.readthedocs.io/en/operational-mission/documentation/mission_module/index.html
@Martin-Delavenne , I asked you as reviewer especially so you can provide feedback on the documentation.

Not addressed in this PR:
- to allow a complete mission specification, the process will need a takeoff segment and the specification of a starting point. Currently, the documentation does not mention these gaps, as I hope to fill them before the 1.0 release.
- A JSON schema validation would be useful to ease file editing. Doing will also question the current validation with `strictYAML` (though strictYAML allows some checks the JSON schema won't)

___

**Additional information:**
A convergence problem on the optimization case showed that the default tolerance (`atol=1e-10`) for NLBGS termination is far too demanding. Setting it to 1e-8 is enough to get back the convergence, and setting it to 1e-2 is enough to keep consistent results (discrepancies are less than 1 gram on MTOW). And of course, relaxing tolerance allows a (slightly) lower CPU consumption.

Martin-Delavenne commented:
Few very general comments about the doc: 

- In the Mission file paragraph, routes section you mention that routes are identified by 4 arguments : "climb_parts", "cruise_part", "descent_parts" and "range". But in the example, when you are describing the diversion_route the argument "parts" appears. If the argument "parts" works with routes keyword shouldn't it be added to the list of admitted arguments? If a route can be alternatively described through the "parts" arguments containing different phases and a cruise segment I think it should also be added in the doc for a greater clarity. 
- In the same example as cited above, the cruise segment is once called with the "cruise" argument the second time with "cruise_type" but this argument is never described later in the doc. 
- In the Segment paragraph, speed_change section, no example is provided.

christophe-david commented:
Thank you @Martin-Delavenne for the comments. My replies below:

> Few very general comments about the doc:
> 
> * In the Mission file paragraph, routes section you mention that routes are identified by 4 arguments : "climb_parts", "cruise_part", "descent_parts" and "range". But in the example, when you are describing the diversion_route the argument "parts" appears. If the argument "parts" works with routes keyword shouldn't it be added to the list of admitted arguments? If a route can be alternatively described through the "parts" arguments containing different phases and a cruise segment I think it should also be added in the doc for a greater clarity.
> * In the same example as cited above, the cruise segment is once called with the "cruise" argument the second time with "cruise_type" but this argument is never described later in the doc.

Actually, the example is wrong. Using `parts` keyword in route definition was my first thought, but since routes are expected to have this particular structure, I considered the definition schema should show it more clearly. Same for `cruise_type`, it was part of the previous syntax and is no more relevant. I will fix the example.

> * In the Segment paragraph, speed_change section, no example is provided.

Indeed. I'll fix that also.

ScottDelbecq reviewed (commented):
Thanks for this great job! Here are my first comments on the docs, my feedback on the code will come soon!

christophe-david commented on docs/documentation/mission_module/mission_file/segments.rst:
```diff
@@ -0,0 +1,316 @@
+Optionally, target altitude can be set to :code:`optimal_flight_level`. In such case, cruise will
+be preceded by a climb segment that will put the aircraft at the altitude that will minimize the
+fuel consumption for the whole segment (including the prepending climb).
+This option is available because the :ref:`segment-altitude_change`segment can reach an altitude
```
Merci. corrigé en 7cb7eaf013dbae5deeb20a4a571b7e97b1c2c3b1.

christophe-david commented on docs/documentation/mission_module/mission_file/segments.rst:
```diff
@@ -0,0 +1,316 @@
+:code:`taxi`
+============
+
+A :code:`taxi`segment simulates the mission parts between gate and takeoff or landing, at constant
```
Merci. corrigé en 7cb7eaf013dbae5deeb20a4a571b7e97b1c2c3b1.

christophe-david commented on docs/documentation/mission_module/mission_module.rst:
```diff
@@ -0,0 +1,169 @@
+Usage in FAST-OAD configuration file
+************************************
+
+The mission module can be used with the identifier :code`"fastoad.performances.mission"`.
```
Et encore merci. corrigé en 7cb7eaf013dbae5deeb20a4a571b7e97b1c2c3b1 :)

christophe-david commented on docs/documentation/mission_module/mission_file/segments.rst:
```diff
@@ -0,0 +1,316 @@
+
+.. _segment-parameter-engine_setting:
+
+:code:`engine_setting`
+======================
+
+Expected value for :code:`engine_setting` are :code:`takeoff`, :code:`climb`
+, :code:`cruise` or :code:`idle`
```
J'ai rajouté un brin d'explication en 0775dc164b8f22e85fd744a66769b17746173358. Dis-moi si ça te semble suffisant.

Mais au fond, c'est sans doute le rubber engine qu'il faudra mieux documenter.

ScottDelbecq replied:
Oui parfait mettre le lien ver rubber engine est très utile.

ScottDelbecq commented on tests/integration_tests/oad_process/test_oad_process.py:
```diff
@@ -13,7 +13,6 @@
 #  GNU General Public License for more details.
 #  You should have received a copy of the GNU General Public License
 #  along with this program.  If not, see <https://www.gnu.org/licenses/>.
-import os
```
`os` is used on line 62.

christophe-david replied:
Fixed in 416455c97505585c058954f152147b57a9168d58

ScottDelbecq reviewed (changes requested):
Thanks for this major work! 
This is a rapid feedback on the usage.
Everything seems to work properly with acceptable computation times except regarding my single comment.

However, I noticed that on the last optimizer iteration in `test_api_optim()` and in the optimization part of the tutorial notebook the MDA fails to converge (even with 300 max. iterations). Since we test the mass breakdown consistency, I guess it is the wing area which is not consistent?

This is not the case when we use the mission file:

```toml
    [model.performance]
        id = "fastoad.performances.mission"
        propulsion_id = "fastoad.wrapper.propulsion.rubber_engine"
        # mission_file_path = "::sizing_breguet"
        mission_file_path = "::sizing_mission"
        out_file = "./flight_points.csv"
        adjust_fuel = true
        is_sizing = true
```

christophe-david commented:
> However, I noticed that on the last optimizer iteration in `test_api_optim()` and in the optimization part of the tutorial notebook the MDA fails to converge (even with 300 max. iterations). Since we test the mass breakdown consistency, I guess it is the wing area which is not consistent?

Thanks for noticing that. I fixed it and updated PR description about that subject.
